scripts/ckan_migration/migrate_harvest_sources.py

```diff
@@ -0,0 +1,219 @@
+logger = logging.getLogger(__name__)
+
+# global config object here so we can set the values from the CLI and then use
+# them in our methods.
+CONFIG = SimpleNamespace()
+
+CONFIG.ckan_harvest_source_query = (
+    "https://catalog.data.gov/api/action/package_search?fq=(dataset_type:harvest)"
+)
+CONFIG.harvester_url = "https://datagov-harvest-admin-dev.app.cloud.gov/"
+
+
+def get_count():
+    """Get the count of harvest sources."""
+    click.echo("Getting number of harvest sources...")
+    data = get(CONFIG.ckan_harvest_source_query + "&rows=0").json()
+    return data["result"]["count"]
+
+
+def get_sources(count):
+    """Get the harvest source information."""
+    click.echo("Getting harvest source information...")
+    data = get(CONFIG.ckan_harvest_source_query + f"&rows={count}&start=0").json()
+    return data["result"]["results"]
+
+
+def _org_to_upload(org_data):
+    return {
+        "id": org_data["id"],
+        "name": org_data["title"],
+        "logo": org_data["image_url"],
+    }
+
+
+def ensure_organization(org_data):
+    """Ensure that the organization from a harvest source exists.
+
+    We won't look before we leap. Try to add the organization and
+    handle any failures.
+
+    Return any organization information we got from queries.
+    """
+    logger.debug(
+        "Ensuring organization %s(%s) exists", org_data["title"], org_data["id"]
+    )
+    upload_data = _org_to_upload(org_data)
+    logger.debug("Creating organization with data %s", upload_data)
+    result = post(
+        CONFIG.harvester_url + "organization/add",
+        json=upload_data,
+        headers=CONFIG.auth_headers,
+    )
+    logger.debug("%s: %s", result.status_code, result.text)
+    if result.status_code == 200:
+        # add succeeded, return the data we used
+        return result.json()
+    else:
+        # add failed, figure out what's going on
+        logger.debug("Failed to add organization...")
+        result = get(
+            CONFIG.harvester_url + f"organization/{upload_data['id']}",
+            headers=CONFIG.auth_headers | {"Content-type": "application/json"},
+        )
+        logger.debug("%s: %s", result.status_code, result.text)
+        return result.json()
+
+
+def _derive_source_fields():
+    """Dict of functions that derive harvester fields from CKAN data.
+
+    The keys of the dict are the field names in Harvester and the values
+    are functions that take the source's CKAN dict and return
+    the value of that field. Mapper functions can take either one argument
+    with just the source information or two for source and organization
+    information.
+    """
+
+    def _source_type(source):
+        if source["source_type"].startswith("waf"):
+            return "waf"
+        return "document"
+
+    def _schema_type(source, org):
+        # TODO: can we actually determine this now?
+        organization_types = [
+            extra["value"]
+            for extra in org.get("extras", [])
```

Comment:
i don't see where you're getting the org extras. am i missing the `organization_show` call?

Reply:
I forgot, adding that now, also putting in the notification_emails which come from that too.